File: LeetCode/DFS/549. Binary Tree Longest Consecutive Sequence II/Solution.dart
import 'dart:math';

class TreeNode {
  int val;
  TreeNode? left;
  TreeNode? right;
  TreeNode([this.val = 0, this.left, this.right]);
}

class Solution {
  int answer = 0;
  int longestConsecutive(TreeNode? root) {
    List<int> dfs(TreeNode? node) {
      if (node == null) return [0, 0];
      int icr = 1, dcr = 1;
      if (node.left != null) {
        List<int> left = dfs(node.left);
        if (node.left?.val == node.val + 1) icr = left[0] + 1;
        if (node.left?.val == node.val - 1) dcr = left[1] + 1;
      }
      if (node.right != null) {
        List<int> right = dfs(node.right);
        if (node.right?.val == node.val + 1) icr = max(icr, right[0] + 1);
        if (node.right?.val == node.val - 1) dcr = max(dcr, right[1] + 1);
      }
      answer = max(answer, icr + dcr - 1);
      return [icr, dcr];
    }

    dfs(root);
    return answer;
  }
}

File: LeetCode/DFS/549. Binary Tree Longest Consecutive Sequence II/Solution.py
class Solution:
    answer = 0

    def longestConsecutive(self, root: TreeNode) -> int:

        def longest_path(node: TreeNode) -> Tuple[int, int]:
            if not node:
                return (0, 0)
            icr = dcr = 1
            if node.left:
                left = longest_path(node.left)
                if node.left.val == node.val+1:
                    icr = left[0]+1
                if node.left.val + 1 == node.val:
                    dcr = left[1]+1

            if node.right:
                right = longest_path(node.right)
                if node.right.val == node.val+1:
                    icr = max(icr, right[0]+1)
                if node.right.val+1 == node.val:
                    dcr = max(dcr, right[1]+1)

            self.answer = max(self.answer, icr+dcr-1)

            return [icr, dcr]
        longest_path(root)
        return self.answer
